x-pack/dockerlogbeat/pipelinemanager/pipelineManager.go

```diff
@@ -55,170 +48,137 @@ func NewPipelineManager(logCfg *common.Config) *PipelineManager {
 
 // CloseClientWithFile closes the client with the associated file
 func (pm *PipelineManager) CloseClientWithFile(file string) error {
-	pm.mu.Lock()
-	defer pm.mu.Unlock()
 
-	cl, ok := pm.clients[file]
-	if !ok {
-		return fmt.Errorf("No client for file %s", file)
+	cl, err := pm.removeClient(file)
+	if err != nil {
+		return errors.Wrap(err, "Error removing client")
 	}
 
-	// deincrement the ref count
 	hash := cl.pipelineHash
-	pm.pipelines[hash].refCount--
 
-	pm.Logger.Infof("Closing Client first from pipelineManager")
-	err := cl.Close()
+	pm.Logger.Debugf("Closing Client first from pipelineManager")
+	err = cl.Close()
 	if err != nil {
 		return errors.Wrap(err, "error closing client")
 	}
 
-	if pm.pipelines[hash].refCount < 1 {
-		pm.Logger.Infof("Pipeline  closing")
-		pm.pipelines[hash].pipeline.Close()
-		delete(pm.pipelines, hash)
-	}
+	//if the pipeline is no longer in use, clean up
+	pm.removePipelineIfNeeded(hash)
 
 	return nil
 }
 
 // CreateClientWithConfig gets the pipeline linked to the given config, and creates a client
 // If no pipeline for that config exists, it creates one.
 func (pm *PipelineManager) CreateClientWithConfig(logOptsConfig map[string]string, file string) (*ClientLogger, error) {
-	pm.mu.Lock()
-	defer pm.mu.Unlock()
 
-	// check to see if we have a pipeline
-	var err error
 	hashstring := makeConfigHash(logOptsConfig)
 
-	err = pm.makePipeline(logOptsConfig, file, hashstring)
-	if err != nil {
-		return nil, err
+	//If we don't have an existing pipeline for this hash, make one
+	exists := pm.checkIfHashExists(logOptsConfig)
+	var pipeline *Pipeline
+	var err error
+	if !exists {
+		pipeline, err = loadNewPipeline(logOptsConfig, file, pm.Logger)
+		if err != nil {
+			return nil, errors.Wrap(err, "error loading pipeline")
+		}
+		pm.registerPipeline(pipeline, hashstring)
```

Comment:
This pattern looks like a race, potentially creating zombie-pipelines. At least the if block should be protected and there should be another check (similar to double-checked locking pattern).

Does docker synchronize API calls on logging plugins? In this case we might be safe here.